xls/contrib/mlir/IR/xls_ops.td

```diff
@@ -305,22 +374,65 @@ class Xls_ArithBinaryOp<string name, string desc>
   );
 }
 
+class Xls_PartialProdOp<string name, string desc> :  Xls_Op<name, [
+  Pure,
+  ShapesAreConsistent<["lhs", "rhs", "result_lhs", "result_rhs"]>,
+  Elementwise,
+  Scalarizable,
+  AllTypesMatch<["result_lhs", "result_rhs"]>
+]> {
+  let summary = !strconcat(desc, " operation");
+  let description = summary;
+  let arguments = (ins
+    Xls_Bits:$lhs,
+    Xls_Bits:$rhs
+  );
+  let results = (outs
+    Xls_Bits:$result_lhs,
+    Xls_Bits:$result_rhs
+  );
+  let assemblyFormat = [{
+    $lhs `,` $rhs attr-dict `:` functional-type(operands, results)
+  }];
+  let builders = [
+    OpBuilder<(ins "::mlir::Type":$result_type, "::mlir::Value":$lhs, "::mlir::Value":$rhs), [{
+      build($_builder, $_state, result_type, result_type, lhs, rhs);
+    }]>,
+  ];
+}
+
 def Xls_AddOp : Xls_ArithBinaryOp<"add", "addition"> {
   let hasFolder = 1;
 }
 def Xls_SmulOp : Xls_ArithBinaryOp<"smul", "signed multiplication">;
+def Xls_SmulpOp : Xls_PartialProdOp<"smulp", [{
+  Signed partial product multiply.
+
+  Returns two results such that:
+    result_lhs + result_rhs = lhs * rhs
+
+  The two results have the same type, but are not fully constrained: Any
+  two values that fulfill the above are valid.
+}]>;
 def Xls_UmulOp : Xls_ArithBinaryOp<"umul", "unsigned multiplication"> {
   let hasFolder = 1;
   let hasCanonicalizeMethod = 1;
 }
+def Xls_UmulpOp : Xls_PartialProdOp<"umulp", [{
+  Unsigned partial product multiply.
+
+  Returns two results such that:
+    result_lhs + result_rhs = lhs * rhs
+
+  The two results have the same type, but are not fully constrained: Any
```

Comment:
Nit: lets make this two sentences. I think the type is constrained but the values aren't. The comma makes me think this is type related.